### ISSUE TITLE
XamlC XC0022 warning : Binding could be compiled if x:DataType is specified after upgrade to 8.0.7

### ISSUE THREAD

**Issue by ChristianRiedl** [dotnet/maui#21288]
### Description

After upgrade to 8.0.7 I get a warning XC0022 and XAML is not compiled.

### Steps to Reproduce

Create a ContentPage like  (line with **ToolbarItem** produces the error) :

```
<ContentPage
    x:Class="MediaApp.Views.EpgPage"
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:vm="clr-namespace:MediaApp.ViewModels"
    x:DataType="vm:EpgViewModel">


    <ContentPage.ToolbarItems>
        <ToolbarItem Text="{Binding PrevText}" IconImageSource="previous.png" >


```...
...


### Link to public reproduction project repository

_No response_

### Version with bug

8.0.7 SR2

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.6 SR1

### Affected platforms

Android, Windows

### Affected platform versions

Windows SDK 10.0.22621.0, Android 34.0

### Did you find any workaround?

Adding x:DataType solved the problem, but this behavior is not as intended I think.

    <ContentPage.ToolbarItems>
        <ToolbarItem x:DataType="vm:EpgViewModel" Text="{Binding PrevText}" IconImageSource="previous.png" >
...

### Relevant log output

```shell
Views\EpgPage.xaml(19,22): XamlC XC0022 warning : Binding could be compiled if x:DataType is specified.
1>Views\EpgPage.xaml(19,75): XamlC XC0022 warning : Binding could be compiled if x:DataType is specified.
1>Views\EpgPage.xaml(20,22): XamlC XC0022 warning : Binding could be compiled if x:DataType is specified.
1>Views\EpgPage.xaml(20,79): XamlC XC0022 warning : Binding could be compiled if x:DataType is specified.
1>Views\EpgPage.xaml(21,22): XamlC XC0022 warning : Binding could be compiled if x:DataType is specified.
...
```

**Comment by PureWeen** [comment]
Duplicate of #20568

**Comment by ChristianRiedl** [comment]
Hi Shane,

I disagree that it is a duplicate of #20568<https://github.com/dotnet/maui/issues/20568>

I specified x:DataType at  ContentPage level. But different to version 8.0.6 this definition is not inherited by the child objects.


<ContentPage
    x:Class="MediaApp.Views.EpgPage"
    xmlns=http://schemas.microsoft.com/dotnet/2021/maui
    xmlns:x=http://schemas.microsoft.com/winfx/2009/xaml
    xmlns:mc=http://schemas.openxmlformats.org/markup-compatibility/2006
    xmlns:vm="clr-namespace:MediaApp.ViewModels"
    x:DataType="vm:EpgViewModel">

    <ContentPage.ToolbarItems>
        <ToolbarItem Text="{Binding PrevText}" IconImageSource="previous.png" Command="{Binding PreviousCommand}"  />

The problem is caused by the ToolBarItem object. A work around is to add x:DataType again in the ToolbarItem.

<ToolbarItem x:DataType="vm:EpgViewModel" Text="{Binding PrevText}" IconImageSource="previous.png" Command="{Binding PreviousCommand}"  />

ToolBarItem does not inherit x:DataType which is not correct in my opinion.

Christian Riedl


Von: Shane Neuville ***@***.***>
Gesendet: Montag, 18. März 2024 21:31
An: dotnet/maui ***@***.***>
Cc: RTS Christian Riedl ***@***.***>; Author ***@***.***>
Betreff: Re: [dotnet/maui] XamlC XC0022 warning : Binding could be compiled if x:DataType is specified after upgrade to 8.0.7 (Issue #21288)


Duplicate of #20568<https://github.com/dotnet/maui/issues/20568>

**Comment by simonrozsival** [comment]
@ChristianRiedl I could not reproduce the issue with the following minimal repro:
```
<ContentPage
    x:Class="Microsoft.Maui.Controls.Xaml.UnitTests.Maui21288"
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:local="clr-namespace:Microsoft.Maui.Controls.Xaml.UnitTests"
    x:DataType="local:ViewModel21288">

    <ContentPage.ToolbarItems>
        <ToolbarItem Text="{Binding PrevText}" />
    </ContentPage.ToolbarItems>
</ContentPage>
```

I'm not getting any warning.

One more thing I noticed is that the warnings that you included refer to lines 19, 20, and 21. None of these lines are shown in the code snippet that you provided. Can you share a snippet that includes those lines?

**Comment by ChristianRiedl** [comment]
I tried to create a simple reproducer, but I was not able. Seems to depend on more things (community toolkit mvvm …)
After upgrading to latest 8.0.10 the problem disappeared. Please close it.

Christian Riedl


Von: dotnet-policy-service[bot] ***@***.***>
Gesendet: Dienstag, 19. März 2024 20:20
An: dotnet/maui ***@***.***>
Cc: RTS Christian Riedl ***@***.***>; Mention ***@***.***>
Betreff: Re: [dotnet/maui] XamlC XC0022 warning : Binding could be compiled if x:DataType is specified after upgrade to 8.0.7 (Issue #21288)


Hi @ChristianRiedl<https://github.com/ChristianRiedl>. We have added the "s/needs-info" label to this issue, which indicates that we have an open question for you before we can take further action. This issue will be closed automatically in 7 days if we do not hear back from you by then - please feel free to re-open it if you come back to this issue after that time.

**Comment by simonrozsival** [comment]
@ChristianRiedl thanks for testing this with the latest release of MAUI. It is likely that the original issue was fixed by https://github.com/dotnet/maui/pull/20742
